### PR TITLE
*: reduce get overlap operations

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #5648, #5606.

### What is changed and how does it work?
```
name                                       old time/op    new time/op    delta
LoadRegionsByVolume/input_size_10000-40      0.05ns ±32%    0.04ns ±18%     ~     (p=0.095 n=5+5)
LoadRegionsByVolume/input_size_100000-40     0.58ns ± 2%    0.46ns ± 4%  -21.66%  (p=0.008 n=5+5)
LoadRegionsByVolume/input_size_1000000-40     6.29s ±11%     4.88s ± 4%  -22.34%  (p=0.008 n=5+5)

name                                       old alloc/op   new alloc/op   delta
LoadRegionsByVolume/input_size_10000-40       0.00B          0.00B          ~     (all equal)
LoadRegionsByVolume/input_size_100000-40      0.00B          0.00B          ~     (all equal)
LoadRegionsByVolume/input_size_1000000-40     934MB ± 0%     933MB ± 0%   -0.01%  (p=0.032 n=5+5)

name                                       old allocs/op  new allocs/op  delta
LoadRegionsByVolume/input_size_10000-40        0.00           0.00          ~     (all equal)
LoadRegionsByVolume/input_size_100000-40       0.00           0.00          ~     (all equal)
LoadRegionsByVolume/input_size_1000000-40     15.2M ± 0%     15.2M ± 0%     ~     (p=0.421 n=5+5)
```

```
name                                        old time/op    new time/op    delta
LoadRegionsByRandomMerge/merge_ratio_0-40      6.51s ± 4%     4.92s ±10%  -24.48%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_20-40     8.06s ± 4%     4.74s ± 8%  -41.23%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_40-40     9.74s ± 3%     4.93s ± 4%  -49.36%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_60-40     11.1s ± 3%      4.9s ± 4%  -55.85%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_80-40     11.9s ± 3%      5.1s ± 7%  -57.32%  (p=0.008 n=5+5)

name                                        old alloc/op   new alloc/op   delta
LoadRegionsByRandomMerge/merge_ratio_0-40      934MB ± 0%     933MB ± 0%     ~     (p=0.056 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_20-40    1.01GB ± 0%    0.93GB ± 0%   -8.26%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_40-40    1.10GB ± 0%    0.93GB ± 0%  -15.72%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_60-40    1.17GB ± 0%    0.93GB ± 0%  -20.13%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_80-40    1.25GB ± 0%    0.95GB ± 0%  -24.07%  (p=0.008 n=5+5)

name                                        old allocs/op  new allocs/op  delta
LoadRegionsByRandomMerge/merge_ratio_0-40      15.2M ± 0%     15.2M ± 0%     ~     (p=0.881 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_20-40     18.0M ± 0%     15.6M ± 0%  -13.49%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_40-40     20.7M ± 0%     15.9M ± 0%  -22.94%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_60-40     23.3M ± 0%     16.3M ± 0%  -30.12%  (p=0.008 n=5+5)
LoadRegionsByRandomMerge/merge_ratio_80-40     25.8M ± 0%     16.6M ± 0%  -35.63%  (p=0.008 n=5+5)
```

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Review by CabinfeverB — COMMENTED** [review]
Pls reslove conflict

File: server/core/region.go [hunk 1]
```diff
@@ -823,6 +842,11 @@ func (r *RegionsInfo) setRegionLocked(region *RegionInfo) (*RegionInfo, []*Regio
 		if rangeChanged {
 			// Delete itself in regionTree so that overlaps will not contain itself.
 			// Because the regionItem is reused, there is no need to delete it in the regionMap.
+			for i, o := range ol {
+				if o.GetID() == region.GetID() {
+					ol = append(ol[:i], ol[i+1:]...)
+				}
+			}
```
Comment (rleungx) [inline server/core/region.go:849]: This part needs to be paid attention to.

**Review by CabinfeverB — APPROVED** [review]

**Review by bufferflies — COMMENTED** [review]
most lgtm

File: server/core/region_tree.go [hunk 2]
```diff
@@ -118,14 +107,17 @@ func (t *regionTree) overlaps(item *regionItem) []*regionItem {
...
-func (t *regionTree) update(item *regionItem) []*RegionInfo {
+func (t *regionTree) update(item *regionItem, withOverlaps bool, overlaps ...*regionItem) []*RegionInfo {
 	region := item.RegionInfo
 	t.totalSize += region.approximateSize
 	regionWriteBytesRate, regionWriteKeysRate := region.GetWriteRate()
 	t.totalWriteBytesRate += regionWriteBytesRate
 	t.totalWriteKeysRate += regionWriteKeysRate
 
-	overlaps := t.overlaps(item)
+	if !withOverlaps {
```
Comment (lhy1024) [inline server/core/region_tree.go:117]: Can we move it out of this function?

**Review by lhy1024 — APPROVED** [review]

File: server/core/region.go [hunk 3]
```diff
@@ -787,29 +815,20 @@ func check(region, origin *RegionInfo, overlaps []*RegionInfo) (*RegionInfo, err
...
-	defer r.t.Unlock()
-	_, overlaps, _ := r.setRegionLocked(region)
-	return overlaps
+	return nil
 }
 
-// SetRegionWithUpdate sets the RegionInfo to regionTree and regionMap and return the update info of subtree.
-func (r *RegionsInfo) SetRegionWithUpdate(region *RegionInfo) (*RegionInfo, []*RegionInfo, bool) {
+// SetRegion sets the RegionInfo to regionTree and regionMap and return the update info of subtree.
+func (r *RegionsInfo) SetRegion(region *RegionInfo, withOverlaps bool, ol ...*regionItem) (*RegionInfo, []*RegionInfo, bool) {
```
Comment (nolouch) [inline server/core/region.go:825]: The function look hard to use....and how to guaranteed overlaps is accurate without lock？

**Comment by rleungx** [comment]
This PR won't be released in v6.5.0.

**Review by nolouch — APPROVED** [review]

**Pull request merged**